File: src/optitrak_stream/scripts/mavros_pose_publisher.py
#!usr/bin/env python3
import rospy
from mavros_msgs.msg import Mavlink
from geometry_msgs.msg import PoseStamped

class PoseSubscriber:
    def __init__(self):
        self.pose_stamped = None
        # Subscriber to a topic
        self.subscriber = rospy.Subscriber("/vrpn_client_node/FASTR/pose", PoseStamped, self.callback)
    
    def callback(self, data):
        # Store the data in the variable
        self.pose_stamped = data
        temp = data.pose.position.x
        # Swapping because Ardupilot X is positive North, but Optitrack y is positive North
        self.pose_stamped.pose.position.x = -1.0 * data.pose.position.y # x is E
        self.pose_stamped.pose.position.y = temp # y is N

    def get_data(self):
        return self.received_data
    
class PosePublisher:
    def __init__(self):
        self.mocap_publisher = rospy.Publisher('/mavros/mocap/pose', PoseStamped, queue_size=10)
        #refresh rate at which information is published
        

    def send_att_pos_mocap(self,pose_stamped):
        # publish the PoseStamped message
        if (pose_stamped != None):
            self.mocap_publisher.publish(pose_stamped)

if __name__ == '__main__':
    rospy.init_node("mavros_pose_publisher")
    rospy.loginfo("optitrak data is streaming")

    pub = PosePublisher()
    sub = PoseSubscriber()

    rate = rospy.Rate(15)

    while not rospy.is_shutdown():
        #TODO stream optitrak data to RPi

        if sub.pose_stamped != None:
            print(f'x = {sub.pose_stamped.pose.position.x:.3f}, y = {sub.pose_stamped.pose.position.y:.3f}, z = {sub.pose_stamped.pose.position.z:.3f}')

        pub.send_att_pos_mocap(sub.pose_stamped)

        # TODO for SLAM later
        # msg = Mavlink()
        # fill in required mavlink msg data

        rate.sleep()

File: src/optitrak_stream/scripts/mavros_pose_publisher.cpp
#include <ros/ros.h>
#include <mavros_msgs/PositionTarget.h>
#include <geometry_msgs/PoseStamped.h>

class PoseSubscriber {
public:
    PoseSubscriber(ros::NodeHandle& nh) {
        // Initialize the subscriber to the /vrpn_client_node/FASTR/pose topic
        subscriber_ = nh.subscribe("/vrpn_client_node/FASTR/pose", 10, &PoseSubscriber::callback, this);
    }

    void callback(const geometry_msgs::PoseStamped::ConstPtr& msg) {
        // Store the received data
        pose_stamped_ = *msg;
    }

    geometry_msgs::PoseStamped getData() {
        return pose_stamped_;
    }

private:
    ros::Subscriber subscriber_;
    geometry_msgs::PoseStamped pose_stamped_;
};

class PosePublisher {
public:
    PosePublisher(ros::NodeHandle& nh) {
        // Initialize the publisher for the /mavros/mocap/pose topic
        mocap_publisher_ = nh.advertise<geometry_msgs::PoseStamped>("/mavros/mocap/pose", 10);
    }

    void sendAttPosMocap(const geometry_msgs::PoseStamped& pose_stamped) {
        // Publish the PoseStamped message
        if (pose_stamped.header.stamp != ros::Time(0)) { // Check if data is available
            mocap_publisher_.publish(pose_stamped);
        }
    }

private:
    ros::Publisher mocap_publisher_;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "mavros_pose_publisher");
    ros::NodeHandle nh;

    ROS_INFO("Optitrack data is streaming");

    PosePublisher pub(nh);
    PoseSubscriber sub(nh);

    ros::Rate rate(15); // 15 Hz

    while (ros::ok()) {
        ros::spinOnce(); // Process incoming messages

        geometry_msgs::PoseStamped pose_stamped = sub.getData();

        // Print received data
        if (pose_stamped.header.stamp != ros::Time(0)) { // Check if data is available
            ROS_INFO("x = %.3f, y = %.3f, z = %.3f",
                     pose_stamped.pose.position.x,
                     pose_stamped.pose.position.y,
                     pose_stamped.pose.position.z);
        }

        pub.sendAttPosMocap(pose_stamped);

        // TODO for SLAM later
        // mavros_msgs::Mavlink msg;
        // fill in required mavlink msg data

        rate.sleep();
    }

    return 0;
}
